src/app/clusters/service-area-server/service-area-cluster-objects.h

```diff
@@ -0,0 +1,377 @@
+     * landmark.
+     * @param[in] aSurfaceTag A common namespace Floor Surface tag - indicates an association of the location with a surface type.
+     *
+     * @note Requirements regarding what combinations of fields and values are valid are not checked by this class.
+     * @note If aLocationName is larger than kLocationNameMaxSize, it will be truncated.
+     * @note If aLocationName is an empty string and aFloorNumber and aAreaTypeTag are null, locationInfo will be set to null.
+     */
+    void Set(uint32_t aLocationId, const DataModel::Nullable<uint8_t> & aMapId, const CharSpan & aLocationName,
+             const DataModel::Nullable<int16_t> & aFloorNumber, const DataModel::Nullable<AreaTypeTag> & aAreaType,
+             const DataModel::Nullable<LandmarkTag> & aLandmarkTag, const DataModel::Nullable<PositionTag> & aPositionTag,
+             const DataModel::Nullable<FloorSurfaceTag> & aSurfaceTag)
+    {
+        locationID = aLocationId;
+        mapID      = aMapId;
+
+        // If there is at least one non-null value for locationInfo, add it to the location structure.
+        if ((!aLocationName.empty()) || (!aFloorNumber.IsNull()) || (!aAreaType.IsNull()))
+        {
+            // Create a home location info structure and fill it in except for the location name. This is done below.
+            locationInfo.locationInfo.SetNonNull(Structs::HomeLocationStruct::Type());
+
+            locationInfo.locationInfo.Value().floorNumber = aFloorNumber;
+            locationInfo.locationInfo.Value().areaType    = aAreaType;
+        }
+        else
+        {
+            locationInfo.locationInfo.SetNull();
+        }
+
+        locationInfo.landmarkTag = aLandmarkTag;
+        locationInfo.positionTag = aPositionTag;
+        locationInfo.surfaceTag  = aSurfaceTag;
+
+        // Save the location name to buffer, so access and lifetime is controlled.
+        // If aLocationName is not null, create a CharSpan to reference the text.
+        // Make sure there is always a terminating character, mainly so we can get c_str for logging.
+        memset(mLocationNameBuffer, 0, sizeof(mLocationNameBuffer));
+
+        // this assumes locationInfo structure was created above, if appropriate
+        if (!locationInfo.locationInfo.IsNull())
+        {
+            if (aLocationName.empty())
+            {
+                locationInfo.locationInfo.Value().locationName = CharSpan(mLocationNameBuffer, 0);
+                mLocationNameBuffer[0]                         = '\0';
+            }
+            else if (aLocationName.size() > sizeof(mLocationNameBuffer))
+            {
+                // Save the truncated name that fits into available size.
+                memcpy(mLocationNameBuffer, aLocationName.data(), sizeof(mLocationNameBuffer));
+                locationInfo.locationInfo.Value().locationName = CharSpan(mLocationNameBuffer, sizeof(mLocationNameBuffer));
+                mLocationNameBuffer[kLocationNameMaxSize]      = '\0';
```

Comment:
So now the CharSpan contains the null?

What should really happen is that things that log should get the CharSpan and then use `%.*s` or so, no?

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/1054f736563033c4f58db3c31fea9d7a26e02df5). 

When printing, we are using the `.data()` method of `CharSpan`. I thought that this was going to return a pointer to the start of the buffer so I was expecting that a `/0` was still required. Apparently not. I tested it.